### PR TITLE
Feature/proof verification

### PR DESCRIPTION
**Pull request by cducrest**
closes: https://github.com/trustlines-protocol/merkle-drop/issues/20

### CODE REVIEW

File: contracts/contracts/MerkleDrop.sol [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+contract MerkleDrop {
+
+    bytes32 public root;
+
+    constructor(bytes32 _root) public {
+        root = _root;
+    }
+
+    function withdraw(address recipient, uint value, bytes32[] memory proof) public {
```
Comment (berndbohmeier) [inline contracts/contracts/MerkleDrop.sol]: we could have 
`withdraw(value, proof)` and `withdrawFor(recipient, value, proof)`
for easier usage
What do you think?

File: contracts/contracts/MerkleDrop.sol [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+        // We need to pack pack the 20 bytes address to the 32 bytes value
+        // to match with the proof made with the python merkle-drop package
+        bytes32 leaf = keccak256(abi.encodePacked(recipient, value));
+        return verifyProof(leaf, proof);
+    }
+
+    function verifyProof(bytes32 leaf, bytes32[] memory proof) internal view returns (bool) {
+        bytes32 currentHash = leaf;
+
+        for (uint i = 0; i <= proof.length - 1; i += 1) {
```
Comment (berndbohmeier) [inline contracts/contracts/MerkleDrop.sol]: this is normally written as `i < proof.length`, please change it to that.

File: merkle_drop/merkle_tree.py [hunk 3]
```diff
@@ -21,6 +21,7 @@ def __init__(
         hash: bytes,
         *,
         parent: "Node" = None,
+        # left and right is arbitrary
```
Comment (berndbohmeier) [inline merkle_drop/merkle_tree.py]: we can rename it to `childs = []` or so if you like that more.

File: contracts/contracts/MerkleDrop.sol (hunk 1 above)
Comment (cducrest) [reply contracts/contracts/MerkleDrop.sol]: Sure.

File: merkle_drop/merkle_tree.py (hunk 3 above)
Comment (cducrest) [reply merkle_drop/merkle_tree.py]: I don't know, I changed it to child1 and child2 but it did not improve it so much so I reverted it and added this comment.

**Pull request merged**